extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.auth.quickconnect;
+
+import java.util.Collection;
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.auth.quickconnect.utility.QCParser;
+import org.apache.guacamole.net.auth.ConnectionGroup;
+import org.apache.guacamole.net.auth.simple.SimpleConnectionDirectory;
+import org.apache.guacamole.net.auth.Connection;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+
+/**
+ * Implementation of the Connection Directory, stored
+ * completely in-memory.
+ */
+public class QuickConnectDirectory extends SimpleConnectionDirectory {
+
+    /**
+     * The unique identifier of the root connection group.
+     */
+    private static final String ROOT_IDENTIFIER = "ROOT";
+
+    /**
+     * The root connection group for this directory.
+     */
+    private final QuickConnectConnectionGroup rootGroup;
+
+    /**
+     * The internal counter for connection IDs.
+     */
+    private int CONNECTION_ID = 0;
+
+    /**
+     * Creates a new QuickConnectDirectory which provides access to the
+     * connections contained within the given Map.
+     *
+     * @param connections
+     *     A Collection of all connections that should be present in this
+     *     connection directory.
+     * @param rootGroup
+     *     A group that should be at the base of this directory.
+     */
+    public QuickConnectDirectory(Collection<Connection> connections, ConnectionGroup rootGroup) {
+        super(connections);
+        this.rootGroup = (QuickConnectConnectionGroup)rootGroup;
+    }
+
+    /**
+     * Returns the current counter and then increments it.
+     *
+     * @return
+     *     An Integer representing the next available connection
+     *     ID to get used when adding connections.
+     */
+    private Integer getNextConnectionID() {
+        return CONNECTION_ID++;
```

Comment:
This is not threadsafe. If multiple calls to `add()` overlap, this value will not be updated as expected. You will need to generate these IDs in a threadsafe manner, or switch to a mechanism that guarantees uniqueness without depending on synchronous updates ([`UUID.randomUUID()`](https://docs.oracle.com/javase/7/docs/api/java/util/UUID.html#randomUUID()) for example).

Reply:
You want to be careful with `UUID.randomUUID()`. The underlying implementation uses `SecureRandom` which can block indefinitely waiting for sufficient entropy (e.g. in the kernel implementation of `/dev/random`). On a system with low activity this can be significant concern. 

`AtomicInteger.getAndIncrement` will give the same semantics as your use of the post-increment operator with thread safety. Performance is generally better than what you can do with your own `ReentrantLock` or (not recommended) an approach using `synchronized`.

Reply:
Not persisting connections at all, and instead dynamically generating the returned `Connection` within `get()` (using the URL as the identifier) would be another option, assuming that persisting the connections in memory is not an intended feature.

Reply:
Went the AtomicInteger route - having the connections persist in-memory is (to me) a desired functionality, so that accomplishes thread-safeness without losing that.  Also cleaned up the code (like my oddly-named all-caps CONNECTION_ID) a little.